File: src/chapters/linear-systems/content/code/gauss-jordan.r
gauss_jordan <- function(A, b) {
  # Solve A x = b by Gauss-Jordan elimination (reduced row echelon form).
  A <- matrix(as.numeric(A), nrow = nrow(A))
  n <- length(b)
  M <- cbind(A, as.numeric(b))
  for (k in 1:n) {
    p <- k - 1 + which.max(abs(M[k:n, k]))   # partial pivot
    if (p != k) M[c(k, p), ] <- M[c(p, k), ]
    M[k, ] <- M[k, ] / M[k, k]               # normalize pivot row
    for (i in 1:n) {
      if (i != k) M[i, ] <- M[i, ] - M[i, k] * M[k, ]   # eliminate
    }
  }
  M[, n + 1]
}

A <- matrix(c( 2,  1, -1,
              -3, -1,  2,
              -2,  1,  2), nrow = 3, byrow = TRUE)
b <- c(8, -11, -3)
cat("x =", gauss_jordan(A, b), "\n")

File: src/chapters/linear-systems/content/code/gauss-jordan.m
function x = gauss_jordan(A, b)
% GAUSS_JORDAN  Solve A x = b via reduced row echelon form.
    n = numel(b); M = [A, b(:)];
    for k = 1:n
        [~, p] = max(abs(M(k:n, k))); p = p + k - 1;
        M([k p], :) = M([p k], :);
        M(k, :) = M(k, :) / M(k, k);                      % normalize pivot row
        for i = 1:n
            if i ~= k
                M(i, :) = M(i, :) - M(i, k) * M(k, :);
            end
        end
    end
    x = M(:, end);
end

% --- Demo ---
A = [2 1 -1; -3 -1 2; -2 1 2]; b = [8; -11; -3];
disp(gauss_jordan(A, b)');
